server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -769,474 +832,503 @@ public void start() {
+            ActionListener.wrap(
+                response -> {
+                    logger.debug("Successfully read cluster state component from remote");
+                    readIndexRoutingTableResults.add(response);
+                },
+                ex -> {
+                    logger.error("Failed to read cluster state from remote", ex);
+                    exceptionList.add(ex);
+                }
+            ),
+            latch
         );
-    }
-
-    BlobPath getManifestFolderPath(String clusterName, String clusterUUID) {
-        return getCusterMetadataBasePath(clusterName, clusterUUID).add(MANIFEST_PATH_TOKEN);
-    }
-
-    /**
-     * Fetch latest index metadata from remote cluster state
-     *
-     * @param clusterUUID             uuid of cluster state to refer to in remote
-     * @param clusterName             name of the cluster
-     * @param clusterMetadataManifest manifest file of cluster
-     * @return {@code Map<String, IndexMetadata>} latest IndexUUID to IndexMetadata map
-     */
-    private Map<String, IndexMetadata> getIndexMetadataMap(
-        String clusterName,
-        String clusterUUID,
-        ClusterMetadataManifest clusterMetadataManifest
-    ) {
-        assert Objects.equals(clusterUUID, clusterMetadataManifest.getClusterUUID())
-            : "Corrupt ClusterMetadataManifest found. Cluster UUID mismatch.";
-        Map<String, IndexMetadata> remoteIndexMetadata = new HashMap<>();
-        for (UploadedIndexMetadata uploadedIndexMetadata : clusterMetadataManifest.getIndices()) {
-            IndexMetadata indexMetadata = getIndexMetadata(clusterName, clusterUUID, uploadedIndexMetadata);
-            remoteIndexMetadata.put(uploadedIndexMetadata.getIndexUUID(), indexMetadata);
-        }
-        return remoteIndexMetadata;
-    }
 
-    /**
-     * Fetch index metadata from remote cluster state
-     *
-     * @param clusterUUID           uuid of cluster state to refer to in remote
-     * @param clusterName           name of the cluster
-     * @param uploadedIndexMetadata {@link UploadedIndexMetadata} contains details about remote location of index metadata
-     * @return {@link IndexMetadata}
-     */
-    private IndexMetadata getIndexMetadata(String clusterName, String clusterUUID, UploadedIndexMetadata uploadedIndexMetadata) {
-        BlobContainer blobContainer = indexMetadataContainer(clusterName, clusterUUID, uploadedIndexMetadata.getIndexUUID());
-        try {
-            String[] splitPath = uploadedIndexMetadata.getUploadedFilename().split("/");
-            return INDEX_METADATA_FORMAT.read(
-                blobContainer,
-                splitPath[splitPath.length - 1],
-                blobStoreRepository.getNamedXContentRegistry()
+        for (UploadedIndexMetadata indexRouting : indicesRoutingToRead) {
+            asyncMetadataReadActions.add(
+                remoteRoutingTableService.get().getAsyncIndexMetadataReadAction(
```

Comment:
Do we need a validation on if the remote routing table is enabled or not here?

Reply:
Added.